functional/stateful-types.md

```diff
@@ -0,0 +1,269 @@
+# Stateful Types: Generics as Type Classes
+
+## Description
+
+Rust's functional roots allow for it to be more expressive in the type system
+than many other languages, and turn many kinds of programming problems into
+"static typing" problems. A key part of this idea is the way generic types work.
+
+In C++ and Java, for example, generic types are a meta-programming construct for
+the compiler. `Vec<int>` and `Vec<char>` in C++ are just two different copies of
+the same boilerplate code for a `Vec` type, with two different types filled in.
+
+In Rust, the generic type parameter creates what is known as a "type class
+constraint", and each value used by an end user *actually changes the type of
+each instatiation*. In other words, `Vec<usize>` and `Vec<char>` *are two
+different types*.
```

Comment:
Isn't this known as **monomorphization**?

Reply:
After thinking about this, I could certainly add that. I was about to ask what C++ calls their template generation -- because I thought I heard a similar word -- but can't find it.